src/sdglib/factories/Kmers.h

```diff
@@ -0,0 +1,181 @@
+        // Here I check someone hasn't requested silly K size, before
+        // making the local translation table for this container.
+        // static asserts below?
+        assert(K > 0);
+        assert(!std::numeric_limits<value_type>::is_signed);
+        // Forgive heavy bracket use - I need to get used to if C++ has the operator precedence I'm used to or not.
+        assert(K <= ((std::numeric_limits<value_type>::digits / 2))); // -1 here to keep space for the sign bits.
+        translation_table = make_nuc2bin_tbl<value_type>();
+    }
+
+    class iterator: public std::iterator<iterator_category, value_type, difference_type, pointer, reference> {
+    public:
+        explicit iterator(Kmers& p) : parent(p) {
+            position = parent.sequence;
+            last_unknown = 0;
+            fwmer = 0;
+            rvmer = 0;
+            mask = (value_type(1) << (2 * parent.K)) - 1;
+            incorporate_char(position);
+        }
+        bool has_reached_end() { return *position == '\0'; }
+
+        reference operator*() const {
+            // I want to specialize this on template parameter F but am unsure on how to do it properly.
+            return std::min(fwmer & mask, rvmer & mask); // SEGFAULTS if I make it non-canonical - WTF.
+        }
+        pointer operator->() const {
+            // I want to specialize this on template parameter F but am unsure on how to do it properly.
+            return fwmer & mask;
+        }
+
+        // Advance the iterator to the next valid kmer in the sequence, or to the end of the string
+        // (\0), whichever comes first.
+        iterator& operator++() {
+            scan_right();
+            return *this;
+        }
+        bool operator<(Kmers::iterator other) {
+            // true if both iterators refer to same sequence, and this iterator is at a lower position.
+            return parent.sequence == other.parent.sequence && position < other.position;
+        }
+        bool operator<=(Kmers::iterator other) {
+            // true if both iterators refer to same sequence, and this iterator is at a lower position.
+            return parent.sequence == other.parent.sequence && position <= other.position;
+        }
+        bool operator!=(Kmers::iterator other) {
+            // true if both iterators refer to the same sequence, and this iterator is not at the same positon.
+            return parent.sequence == other.parent.sequence && position != other.position;
+        }
+        bool operator==(Kmers::iterator other) {
+            // true if both iterators refer to the same sequence, and this iterator is not at the same positon.
+            return parent.sequence == other.parent.sequence && position == other.position;
+        }
+    protected:
+        const Kmers& parent;
+        const char* position;
+        value_type fwmer;
+        value_type rvmer;
+        value_type mask;
+        size_t last_unknown;
```

Comment:
the name of this variable is wrong. It may be my fault on the original code, or history as we optimised this. but the variable actually holds "characters since last unknown" or "valid_nucleotides".